// @procore/core-react/dist/esm/components/Thumbnail/Thumbnail.styles.js
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
import styled, { css } from 'styled-components';
import { borderRadius } from '../../styles/borderRadius';
import { colors } from '../../styles/colors';
import { getEllipsis, getFocus } from '../../styles/mixins';
import { spacing } from '../../styles/spacing';
import { Typography } from '../Typography';
import { DocumentIcon } from './DocumentIcon';
var thumbnailSizeSm = 40;
var thumbnailSizeLg = 104;
var labelHeight = 24;
var labelFontSize = 12;
var tickSize = 24;
var rowLayoutCaptionWidth = 288;
var multilineCaptionMaxHeight = spacing.lg * 2;
var captionBase = css(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  color: ", ";\n  font-size: 12px;\n  line-height: 16px;\n  letter-spacing: 0.4px;\n  font-weight: 500;\n  ", "\n  ", "\n\n  ", "\n  ", "\n"], ["\n  color: ", ";\n  font-size: 12px;\n  line-height: 16px;\n  letter-spacing: 0.4px;\n  font-weight: 500;\n  ", "\n  ", "\n\n  ", "\n  ", "\n"])), colors.gray15, function (_a) {
    var layout = _a.layout;
    return layout === 'row'
        ? "\n        align-items: center;\n        margin-left: ".concat(spacing.sm, "px;\n        width: ").concat(rowLayoutCaptionWidth, "px;\n      ")
        : '';
}, function (_a) {
    var layout = _a.layout;
    return layout === 'column'
        ? "\n        margin-top: ".concat(spacing.xs / 2, "px;\n        width: ").concat(thumbnailSizeLg, "px;\n      ")
        : '';
}, function (_a) {
    var error = _a.error;
    return (error ? "color: ".concat(colors.red50, ";") : '');
}, function (_a) {
    var disabled = _a.disabled;
    return (disabled ? "color: ".concat(colors.gray70, ";") : '');
});
export var StyledMultilineCaption = styled.div(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  ", "\n  display: flex;\n  word-wrap: break-word;\n  // for IE11, use word-wrap property\n  overflow-wrap: break-word;\n  overflow: hidden;\n  display: -webkit-box;\n  -webkit-line-clamp: 2;\n  -webkit-box-orient: vertical;\n\n  // limit height for browsers that do not support line-clamp CSS prop\n  max-height: ", "px;\n"], ["\n  ", "\n  display: flex;\n  word-wrap: break-word;\n  // for IE11, use word-wrap property\n  overflow-wrap: break-word;\n  overflow: hidden;\n  display: -webkit-box;\n  -webkit-line-clamp: 2;\n  -webkit-box-orient: vertical;\n\n  // limit height for browsers that do not support line-clamp CSS prop\n  max-height: ", "px;\n"])), captionBase, multilineCaptionMaxHeight);
export var StyledCaptionText = styled.span(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  // fix word wrap issue in IE\n  width: 100%;\n"], ["\n  // fix word wrap issue in IE\n  width: 100%;\n"])));
export var StyledPlaceholderCaption = styled.div(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  ", "\n"], ["\n  ", "\n"])), captionBase);
export var StyledSingleLineCaption = styled.div(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n  display: flex;\n  ", "\n"], ["\n  display: flex;\n  ", "\n"])), captionBase);
export var StyledSinglelineCaptionBody = styled.span(templateObject_6 || (templateObject_6 = __makeTemplateObject(["\n  ", "\n"], ["\n  ", "\n"])), getEllipsis);
export var StyledSinglelineCaptionEnding = styled.span(templateObject_7 || (templateObject_7 = __makeTemplateObject(["\n  max-width: 50%;\n  white-space: nowrap;\n  flex-shrink: 0;\n"], ["\n  max-width: 50%;\n  white-space: nowrap;\n  flex-shrink: 0;\n"])));
export var StyledLabelText = styled(Typography)(templateObject_8 || (templateObject_8 = __makeTemplateObject(["\n  ", "\n"], ["\n  ", "\n"])), getEllipsis());
export var StyledLabel = styled.div(templateObject_9 || (templateObject_9 = __makeTemplateObject(["\n  display: flex;\n  justify-content: center;\n  position: absolute;\n  bottom: 0;\n  right: 0;\n  left: 0;\n  opacity: 0.85;\n  height: ", "px;\n  padding: ", "px ", "px;\n  font-size: ", "px;\n  background-color: ", ";\n"], ["\n  display: flex;\n  justify-content: center;\n  position: absolute;\n  bottom: 0;\n  right: 0;\n  left: 0;\n  opacity: 0.85;\n  height: ", "px;\n  padding: ", "px ", "px;\n  font-size: ", "px;\n  background-color: ", ";\n"])), labelHeight, spacing.sm / 2, spacing.sm, labelFontSize, colors.white);
var thumbnailBase = css(templateObject_10 || (templateObject_10 = __makeTemplateObject(["\n  display: flex;\n  justify-content: center;\n  align-items: center;\n  position: relative;\n  overflow: hidden;\n  border: 1px solid ", ";\n  border-radius: ", "px;\n\n  ", "\n\n  ", "\n\n  &:focus {\n    ", ";\n  }\n\n  ", "\n\n  ", "\n\n  ", "\n"], ["\n  display: flex;\n  justify-content: center;\n  align-items: center;\n  position: relative;\n  overflow: hidden;\n  border: 1px solid ", ";\n  border-radius: ", "px;\n\n  ", "\n\n  ", "\n\n  &:focus {\n    ", ";\n  }\n\n  ", "\n\n  ", "\n\n  ", "\n"])), colors.gray85, borderRadius.md, function (_a) {
    var clickable = _a.clickable;
    return (clickable ? 'cursor: pointer;' : '');
}, function (_a) {
    var selected = _a.selected;
    return selected
        ? "\n        ".concat(StyledTickIcon, " {\n          opacity: 1;\n          transform: scale(1);\n        }\n      ")
        : '';
}, getFocus(), function (_a) {
    var focused = _a.focused;
    return focused && getFocus();
}, function (_a) {
    var disabled = _a.disabled, clickable = _a.clickable;
    return disabled && clickable ? 'cursor: not-allowed;' : '';
}, function (_a) {
    var size = _a.size;
    var sizeInPx = "".concat(size === 'lg' ? thumbnailSizeLg : thumbnailSizeSm, "px");
    return "\n      width: ".concat(sizeInPx, ";\n      height: ").concat(sizeInPx, ";\n    ");
});
export var StyledTickIcon = styled.svg(templateObject_11 || (templateObject_11 = __makeTemplateObject(["\n  position: absolute;\n  top: ", "px;\n  right: ", "px;\n  width: ", "px;\n  height: ", "px;\n  opacity: 0;\n  transform: scale(0.75);\n  transition: transform 250ms ease-out, opacity 100ms ease-out;\n\n  ", "\n\n  ", "\n"], ["\n  position: absolute;\n  top: ", "px;\n  right: ", "px;\n  width: ", "px;\n  height: ", "px;\n  opacity: 0;\n  transform: scale(0.75);\n  transition: transform 250ms ease-out, opacity 100ms ease-out;\n\n  ", "\n\n  ", "\n"])), spacing.sm, spacing.sm, tickSize, tickSize, function (_a) {
    var selected = _a.selected;
    return selected
        ? "\n        opacity: 1;\n        transform: scale(1);\n      "
        : '';
}, function (_a) {
    var size = _a.size;
    return size === 'sm'
        ? "\n        top: 50%;\n        left: 50%;\n        margin-top: ".concat(-tickSize / 2, "px;\n        margin-left: ").concat(-tickSize / 2, "px;\n      ")
        : '';
});
export var StyledFileIcon = styled(DocumentIcon)(templateObject_12 || (templateObject_12 = __makeTemplateObject(["\n  width: 36.5%;\n  height: auto;\n"], ["\n  width: 36.5%;\n  height: auto;\n"])));
export var StyledImageThumbnailPreview = styled.div(templateObject_14 || (templateObject_14 = __makeTemplateObject(["\n  width: 100%;\n  height: 100%;\n  background-size: cover;\n  background-repeat: no-repeat;\n  background-position: center;\n  ", "\n"], ["\n  width: 100%;\n  height: 100%;\n  background-size: cover;\n  background-repeat: no-repeat;\n  background-position: center;\n  ", "\n"])), function (_a) {
    var $bgImage = _a.$bgImage;
    return css(templateObject_13 || (templateObject_13 = __makeTemplateObject(["\n    background-image: url(", ");\n  "], ["\n    background-image: url(", ");\n  "])), $bgImage);
});
export var StyledImageThumbnailOverlay = styled.div(templateObject_15 || (templateObject_15 = __makeTemplateObject(["\n  display: none;\n  position: absolute;\n  top: 0;\n  bottom: 0;\n  left: 0;\n  right: 0;\n"], ["\n  display: none;\n  position: absolute;\n  top: 0;\n  bottom: 0;\n  left: 0;\n  right: 0;\n"])));
var blackOverlay = "\n  background-color: ".concat(colors.black, ";\n  opacity: 0.2;\n");
var whiteOverlay = "\n  background-color: ".concat(colors.white, ";\n  opacity: 0.5;\n");
export var StyledImageThumbnailImage = styled.img(templateObject_16 || (templateObject_16 = __makeTemplateObject([""], [""])));
export var StyledImageThumbnail = styled.div(templateObject_17 || (templateObject_17 = __makeTemplateObject(["\n  ", "\n\n  ", "\n\n  ", "\n\n  ", "\n\n  ", "\n\n  ", "\n\n  ", "\n"], ["\n  ", "\n\n  ", "\n\n  ", "\n\n  ", "\n\n  ", "\n\n  ", "\n\n  ", "\n"])), thumbnailBase, function (_a) {
    var disabled = _a.disabled;
    return disabled
        ? "\n        &,\n        &:hover {\n          ".concat(StyledImageThumbnailOverlay, " {\n            display: block;\n            ").concat(whiteOverlay, "\n          }\n        }\n        ").concat(StyledImageThumbnailPreview, ",\n        ").concat(StyledTickIcon, ",\n        ").concat(StyledImageThumbnailImage, " {\n          filter: saturate(0);\n        }\n      ")
        : '';
}, function (_a) {
    var selected = _a.selected, focused = _a.focused, disabled = _a.disabled;
    return focused && disabled
        ? selected
            ? "\n          ".concat(StyledTickIcon, " {\n            z-index: 0;\n          }\n        ")
            : "\n          ".concat(StyledTickIcon, " {\n            z-index: 1;\n          }\n        ")
        : '';
}, function (_a) {
    var selected = _a.selected, focused = _a.focused;
    return selected && focused
        ? "\n        ".concat(StyledTickIcon, " {\n          z-index: 1;\n        }\n      ")
        : '';
}, function (_a) {
    var selected = _a.selected, focused = _a.focused, disabled = _a.disabled;
    return selected && focused && disabled
        ? "\n        ".concat(StyledTickIcon, " {\n          z-index: 0;\n        }\n      ")
        : '';
}, function (_a) {
    var focused = _a.focused, disabled = _a.disabled;
    return focused && !disabled
        ? "\n        ".concat(StyledImageThumbnailOverlay, " {\n          display: block;\n          ").concat(blackOverlay, "\n        }\n      ")
        : '';
}, function (_a) {
    var clickable = _a.clickable, disabled = _a.disabled;
    return clickable && !disabled
        ? "\n        &:hover {\n          ".concat(StyledImageThumbnailOverlay, " {\n            display: block;\n            ").concat(blackOverlay, "\n          }\n        }\n      ")
        : '';
});
export var StyledThumbnailWrapper = styled.div(templateObject_18 || (templateObject_18 = __makeTemplateObject(["\n  display: flex;\n  ", "\n  ", "\n"], ["\n  display: flex;\n  ", "\n  ", "\n"])), function (_a) {
    var size = _a.size;
    return size === 'lg'
        ? "\n        width: ".concat(thumbnailSizeLg, "px;\n      ")
        : '';
}, function (_a) {
    var layout = _a.layout;
    var isRowLayout = layout === 'row';
    return isRowLayout
        ? "\n        flex-direction: row;\n        aligh-items: center;\n      "
        : "\n        flex-direction: column;\n        aligh-items: flex-start;\n      ";
});
export var StyledPlaceholderThumbnail = styled.div(templateObject_19 || (templateObject_19 = __makeTemplateObject(["\n  ", "\n\n  ", "\n\n  ", "\n\n  ", "\n\n  ", "\n"], ["\n  ", "\n\n  ", "\n\n  ", "\n\n  ", "\n\n  ", "\n"])), thumbnailBase, function (_a) {
    var clickable = _a.clickable;
    return clickable
        ? "\n        &:hover {\n          background-color: ".concat(colors.gray96, ";\n        }\n\n      ")
        : '';
}, function (_a) {
    var selected = _a.selected, size = _a.size;
    return selected && size === 'sm'
        ? "\n        ".concat(StyledFileIcon, " {\n          display: none;\n        }\n      ")
        : '';
}, function (_a) {
    var disabled = _a.disabled;
    return disabled
        ? "\n        ".concat(StyledFileIcon, ",\n        ").concat(StyledTickIcon, " {\n          filter: saturate(0);\n          opacity: 0.5;\n        }\n\n        &:hover {\n          background-color: inherit;\n        }\n      ")
        : '';
}, function (_a) {
    var disabled = _a.disabled, selected = _a.selected;
    return disabled && !selected
        ? "\n        ".concat(StyledTickIcon, " {\n          opacity: 0;\n          transform: scale(0.75);\n        }\n      ")
        : '';
});
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7, templateObject_8, templateObject_9, templateObject_10, templateObject_11, templateObject_12, templateObject_13, templateObject_14, templateObject_15, templateObject_16, templateObject_17, templateObject_18, templateObject_19;
