// @procore/core-react/dist/esm/components/FileSelect/ThumbnailList/ThumbnailList.styles.js
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
import { FixedSizeList as List } from 'react-window';
import styled from 'styled-components';
import { colors } from '../../../styles/colors';
import { spacing } from '../../../styles/spacing';
import { Icon } from '../../Icon';
import { Thumbnail } from '../../Thumbnail';
import { getTypographyIntent } from '../../Typography/Typography.styles';
export var StyledThumbnailListContainer = styled.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  display: flex;\n  flex-direction: column;\n  align-items: stretch;\n  outline: none;\n"], ["\n  display: flex;\n  flex-direction: column;\n  align-items: stretch;\n  outline: none;\n"])));
export var StyledThumbnailList = styled(List)(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  outline: none;\n"], ["\n  outline: none;\n"])));
export var StyledThumbnail = styled(Thumbnail)(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  flex-grow: 1;\n  justify-content: flex-start;\n"], ["\n  flex-grow: 1;\n  justify-content: flex-start;\n"])));
export var StyledProgress = styled.div(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  width: 146px;\n  padding: 0 ", "px 0 ", "px;\n  margin-left: auto;\n"], ["\n  width: 146px;\n  padding: 0 ", "px 0 ", "px;\n  margin-left: auto;\n"])), spacing.lg, spacing.xl);
export var StyledThumbnailListItem = styled.div(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n  display: flex;\n  flex-grow: 1;\n  align-items: center;\n  padding: ", "px ", "px ", "px ", "px;\n  ", "\n\n  ", "\n\n  ", "\n\n  ", "\n\n  ", "\n"], ["\n  display: flex;\n  flex-grow: 1;\n  align-items: center;\n  padding: ", "px ", "px ", "px ", "px;\n  ", "\n\n  ", "\n\n  ", "\n\n  ", "\n\n  ", "\n"])), spacing.sm, spacing.lg, spacing.sm, spacing.sm, getTypographyIntent('body'), function (_a) {
    var focused = _a.focused;
    return focused &&
        "\n      ".concat(StyledDeleteIcon, " {\n        color: ").concat(colors.white, ";\n\n        background-color: ").concat(colors.gray50, ";\n        border-radius: 50%;\n\n        path {\n          fill: ").concat(colors.white, ";\n        }\n      }\n    ");
}, function (_a) {
    var error = _a.error;
    return !error &&
        "\n      &:hover {\n        background-color: ".concat(colors.gray96, ";\n      }\n    ");
}, function (_a) {
    var focused = _a.focused, error = _a.error;
    return focused &&
        !error &&
        "\n      background-color: ".concat(colors.gray96, ";\n    ");
}, function (_a) {
    var error = _a.error;
    return error &&
        "\n      background-color: ".concat(colors.red98, ";\n\n      ").concat(StyledProgress, " {\n        color: ").concat(colors.red50, ";\n      }\n    ");
});
export var StyledDeleteIcon = styled(Icon)(templateObject_6 || (templateObject_6 = __makeTemplateObject([""], [""])));
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6;
