// @procore/core-react/dist/esm/components/_storyHelpers_/constants.js
export var exampleText = {
    default: 'Click me!',
    none: '',
    short_ipsum: 'Lorum',
    short_sentence: 'Did you think you have mastery over fire? Imbecile!',
    long_ipsum: "Lorem ip sum dolor sit amet, consectetur adipiscing elit,\n    ed do eiusmod tempor incididunt ut labore et dolore magna aliqua.\n    Scelerisque eleifend donec pretium vulputate sapien nec sagittis.\n    Ultricies tristique nulla aliquet enim tortor at auctor urna nunc.\n    Vitae turpis massa sed elementum tempus egestas.\n    Nunc consequat interdum varius sit amet",
    longest_ipsum: "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod\n    tempor incididunt ut labore et dolore magna aliqua. Ac tortor dignissim\n    convallis aenean et. Facilisis volutpat est velit egestas dui id. In\n    nulla posuere sollicitudin aliquam ultrices. Lorem mollis aliquam ut\n    porttitor. Convallis aenean et tortor at risus viverra adipiscing at.\n    Euismod nisi porta lorem mollis aliquam ut porttitor. Hac habitasse\n    platea dictumst vestibulum rhoncus est pellentesque elit ullamcorper.\n    Viverra tellus in hac habitasse platea dictumst. Scelerisque viverra lawn\n    gnome in aliquam sem fringilla ut morbi tincidunt. Ullamcorper morbi tincidunt\n    ornare massa eget. Gravida cum sociis natoque penatibus. Leo vel fringilla est\n    ullamcorper eget nulla. Vitae proin sagittis nisl rhoncus mattis rhoncus urna.\n    Mauris commodo quis imperdiet massa tincidunt nunc pulvinar sapien et. Id aliquet\n    lectus proin nibh troll condimentum id venenatis.",
    long_ipsum_one_line: "Lorem ip sum dolor sit amet, consectetur adipiscing elit, ed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Scelerisque eleifend donec pretium vulputate sapien nec sagittis. Ultricies tristique nulla aliquet enim tortor at auctor urna nunc. Vitae turpis massa sed elementum tempus egestas. Nunc consequat interdum varius sit amet",
    number: '0123456789',
    symbol: ')!@#$%^&*(',
    no_space: 'ThequickbrownfoxjumpedoverthelazydogItwasindeedthequickbrownfoxthatjumpedoverthelazydog',
    intl_symbol: '¡¿ßøÆ«»ºª',
    mixed_char: 'a!2#4%6 8»',
    alert_text: 'By fire be purged!',
};
export function Ipsum(lengthOrLengthProp) {
    return recursiveGrow(typeof lengthOrLengthProp === 'number'
        ? lengthOrLengthProp
        : lengthOrLengthProp.length);
}
var recursiveGrow = function (length, out) {
    if (length === void 0) { length = 1; }
    if (out === void 0) { out = exampleText.longest_ipsum; }
    if (!length) {
        return '';
    }
    if (out.length < length) {
        return recursiveGrow(length, out + exampleText.longest_ipsum);
    }
    if (out.length >= length) {
        return out.slice(0, length - 1) + '.';
    }
    return 'Lorem ipsum grow failed.';
};
export var commonMimeTypes = {
    '.aac': 'audio/aac',
    '.abw': 'application/x-abiword',
    '.arc': 'application/x-freearc',
    '.avi': 'video/x-msvideo',
    '.azw': 'application/vnd.amazon.ebook',
    '.bin': 'application/octet-stream',
    '.bmp': 'image/bmp',
    '.bz': 'application/x-bzip',
    '.bz2': 'application/x-bzip2',
    '.csh': 'application/x-csh',
    '.css': 'text/css',
    '.csv': 'text/csv',
    '.doc': 'application/msword',
    '.docx': 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
    '.eot': 'application/vnd.ms-fontobject',
    '.epub': 'application/epub+zip',
    '.gz': 'application/gzip',
    '.gif': 'image/gif',
    '.htm': 'text/html',
    '.html': 'text/html',
    '.ico': 'image/vnd.microsoft.icon',
    '.ics': 'text/calendar',
    '.jar': 'application/java-archive',
    '.jpeg': 'image/jpeg',
    '.jpg': 'image/jpeg',
    '.js': 'text/javascript',
    '.json': 'application/json',
    '.jsonld': 'application/ld+json',
    '.mid': 'audio/midi',
    '.midi': 'audio/x-midi',
    '.mjs': 'text/javascript',
    '.mp3': 'audio/mpeg',
    '.mpeg': 'video/mpeg',
    '.mpkg': 'application/vnd.apple.installer+xml',
    '.odp': 'application/vnd.oasis.opendocument.presentation',
    '.ods': 'application/vnd.oasis.opendocument.spreadsheet',
    '.odt': 'application/vnd.oasis.opendocument.text',
    '.oga': 'audio/ogg',
    '.ogv': 'video/ogg',
    '.ogx': 'application/ogg',
    '.opus': 'audio/opus',
    '.otf': 'font/otf',
    '.png': 'image/png',
    '.pdf': 'application/pdf',
    '.php': 'application/php',
    '.ppt': 'application/vnd.ms-powerpoint',
    '.pptx': 'application/vnd.openxmlformats-officedocument.presentationml.presentation',
    '.rar': 'application/vnd.rar',
    '.rtf': 'application/rtf',
    '.sh': 'application/x-sh',
    '.svg': 'image/svg+xml',
    '.swf': 'application/x-shockwave-flash',
    '.tar': 'application/x-tar',
    '.tif': 'image/tiff',
    '.tiff': 'image/tiff',
    '.ts': 'video/mp2t',
    '.ttf': 'font/ttf',
    '.txt': 'text/plain',
    '.vsd': 'application/vnd.visio',
    '.wav': 'audio/wav',
    '.weba': 'audio/webm',
    '.webm': 'video/webm',
    '.webp': 'image/webp',
    '.woff': 'font/woff',
    '.woff2': 'font/woff2',
    '.xhtml': 'application/xhtml+xml',
    '.xls': 'application/vnd.ms-excel',
    '.xlsx': 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
    '.xml': 'application/xml, text/xml',
    '.xul': 'application/vnd.mozilla.xul+xml',
    '.zip': 'application/zip',
    '.3gp': 'video/3gpp',
    '.3gp ': 'audio/3gpp',
    '.3g2': 'video/3gpp2, audio/3gpp2',
    '.7z': 'application/x-7z-compressed',
};
export var mockedAcceptFileTypes = [
    {
        id: 1,
        label: 'image/*',
        mimetypes: ['image/*'],
    },
    {
        id: 2,
        label: 'audio/*',
        mimetypes: ['audio/*'],
    },
    {
        id: 3,
        label: 'video/*',
        mimetypes: ['video/*'],
    },
    {
        id: 4,
        label: '.pdf',
        mimetypes: ['application/pdf'],
    },
    {
        id: 5,
        label: '.doc, .docx',
        mimetypes: [
            'application/msword',
            'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
        ],
    },
    {
        id: 6,
        label: '.ppt, .pptx',
        mimetypes: [
            'application/vnd.ms-powerpoint',
            'application/vnd.openxmlformats-officedocument.presentationml.presentation',
        ],
    },
];
export var avatarImageUrl = 'https://en.gravatar.com/userimage/46725293/c50eaab013a795dae218cbc7e91227eb.jpeg';
// https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
export var tzDatabaseNames = [
    'Pacific/Honolulu',
    'America/Los_Angeles',
    'America/Chicago',
    'America/Costa_Rica',
    'America/New_York',
    'Europe/London',
    'Africa/Maputo',
    'Europe/Kiev',
    'Asia/Dubai',
    'Australia/Perth',
    'Asia/Seoul',
    'Australia/Sydney',
    'Pacific/Auckland',
];
export var locales = [
    'en',
    'en-CA',
    'fr-CA',
    'fr-FR',
    'es',
    'es-ES',
    'en-AE',
    'en-AU',
    'en-GB',
    'en-SG',
    'ko',
    'th-TH',
    'pt',
    'is-IS',
    'de-DE',
    'pseudo',
];
