// @procore/core-react/.storybook/main.js
const ie11Rule = {
  test: /\.(js)$/,
  exclude: /@babel(?:\/|\\{1,2})runtime|core-js/,
  use: [
    {
      loader: 'babel-loader',
      options: {
        presets: [
          [
            '@babel/preset-env',
            {
              useBuiltIns: 'entry',
              targets: {
                ie: '11',
              },
              corejs: '2.6',
            },
          ],
        ],
      },
    },
  ],
}

module.exports = {
  webpackFinal: async (config, { configType }) => {
    // `configType` has a value of 'DEVELOPMENT' or 'PRODUCTION'
    // You can change the configuration based on that.
    // 'PRODUCTION' is used when building the static version of storybook.

    config.resolve.extensions = ['.ts', '.tsx', '.js', '.css', '.scss']
    config.module.rules.push({
      test: /\.scss$/,
      use: [
        'style-loader',
        {
          loader: 'css-loader',
          options: {
            esModule: false,
            modules: {
              localIdentName: 'core-[local]',
            },
          },
        },
        'sass-loader',
      ],
    })
    config.module.rules.push(ie11Rule)

    return config
  },
  typescript: {
    checkOptions: {
      check: true,
      exclude: /node_modules/,
    },
    reactDocgen: false,
  },
  stories: ['../src/**/*.stories.@(js|ts|jsx|tsx)'],
  addons: [
    '@storybook/addon-essentials',
    '@storybook/addon-storysource',
    '@storybook/addon-a11y',
  ],
}
